### PR TITLE
feat(lsp): add API for workspace file operations

### PR DESCRIPTION
**Pull request by stevearc**
related to #20784

This adds support for the workspace file operations: willRenameFiles, willCreateFiles, willDeleteFiles, didRenameFiles, didCreateFiles, and didDeleteFiles.

This will not change the default Neovim behavior at all. It is intended for file management plugins such as nvim-tree or oil.nvim to call these new methods when they perform file operations.

Some of the notable design decisions:
* The API is synchronous. From [the specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_willRenameFiles): "Please note that clients might drop results if computing the edit took too long or if a server constantly fails on this request. This is done to keep renames fast and reliable." We can take this to mean that clients are expected to block the file operation until the request completes, so a sync method makes sense. If we want to make it async later, it can be done by adding an optional `callback` argument and an `async = true` field to the options table.
* The `will*` methods return the list of workspace edits. This allows the caller to see what edits were performed as a result of the file operations. Combined with `apply_edits = false`, this allows the caller to do custom handling of the edits (potentially prompting the user for confirmation before applying). Note that because `vim.lsp.util.apply_workspace_edit` requires the offset encoding, we have to also return the offset encoding as well.
* I had to add a new `vim.fs.path_contains` method to check when file paths are in a LSP workspace. Open to suggestions for different names, or if we don't want to commit to exposing that via the public API for now, I can make it a local function.
* The methods were added to a new lua module `vim.lsp.workspace`. Other similar methods seem to live in `vim.lsp.buf`, but it really doesn't make sense for workspace methods to be under the `buf` namespace.
* The logic for using vim.lpeg to create a match pattern from a glob is moved into `vim.lsp.util`. There are several places in the LSP spec that use globs, and it's very useful to have that as a generally-available utility.

TODO:
- [ ] Tests for path_contains
- [ ] Regenerate documentation?
- [ ] Tests for vim.lsp.util.glob_to_lpeg
- [ ] Tests for vim.lsp.workspace.*

### CODE REVIEW

File: runtime/lua/vim/lsp/util.lua [hunk 1]
```diff
@@ -544,6 +545,74 @@ end
 -- function M.glob_to_regex(glob)
 -- end
 
+--- Parses the raw glob into an |lpeg| pattern. LPeg patterns natively support the "this" or "that"
+--- alternative constructions described in the LSP spec that cannot be expressed in a standard Lua pattern.
+---
+---@param pattern string The raw glob pattern
+---@return vim.lpeg.Pattern? pattern An |lpeg| representation of the pattern, or nil if the pattern is invalid.
+function M.glob_to_lpeg(pattern)
```
Comment (justinmk) [inline runtime/lua/vim/lsp/util.lua:553]: this change also would be helpful as a separate PR, then we can focus discussion for the workspace changes

File: runtime/lua/vim/fs.lua [hunk 2]
```diff
@@ -367,4 +367,33 @@ function M.normalize(path, opts)
...
 end
 
+--- Check if a path is contained within another path.
+---@param container string The path being checked to see if it is the parent
+---@param path string The path being checked to see if it is contained within the parent
+---@return boolean
+function M.path_contains(container, path)
+  -- Normalize both paths so we don't have to deal with different path separators
+  container = M.normalize(vim.fn.fnamemodify(container, ':p'))
+  path = M.normalize(vim.fn.fnamemodify(path, ':p'))
```
Comment (mike325) [inline runtime/lua/vim/fs.lua:377]: `libuv` has `fs_realpath`, it may be better to use libuv's functions instead of `vim.fn.*` to allow the method to be used in fast event context and leave the door open for users to use them in threads

Comment (clason) [reply runtime/lua/vim/fs.lua:377]: If we don't use libuv in a fairly recent API, there's a reason for it. Vim's function are (sadly, for libuv), more robust and platform-independent especially around edge cases.

File: runtime/lua/vim/lsp/util.lua (hunk 1 above)
Comment (mfussenegger) [reply runtime/lua/vim/lsp/util.lua:553]: +1

And it probably shouldn't go into lsp.util, but another location. See https://github.com/neovim/neovim/issues/23291#issuecomment-1817873088 and https://github.com/neovim/neovim/issues/25272

And it would be nice if we could also document the supported glob pattern.

> Tests for vim.lsp.util.glob_to_lpeg

The `_watchfiles_spec` is currently mostly testing the pattern matching, we can move these too.

**Review by mfussenegger — COMMENTED** [review]
I'm not really sure about this approach.

I didn't really think this through yet, but I wonder if it would make more sense to instead expose some high level functions. E.g. Something like `vim.fs.rename(old, new)` and `vim.fs.delete(path)`.

These would either invoke the lsp actions directly (if we're okay coupling that), or we introduce new autocmd events like `FileRename` and `FileDelete` and LSP subscribes to them.


The events would allow more flexible composition and the higher level APIs might be more useful to users.

Exposing APIs that are pretty much only there to make things a bit more convenient for plugin authors seems a bit limiting.


---

We also need to check how these are supposed to interact with existing creation, rename and deletion activities, like when we apply workspace edits:

https://github.com/neovim/neovim/blob/529498685bbcd4783bc0e816d6247118c9ffb9a7/runtime/lua/vim/lsp/util.lua#L758-L767

**Comment by justinmk** [comment]
> E.g. Something like `vim.fs.rename(old, new)` and `vim.fs.delete(path)`.
> 
> These would either invoke the lsp actions directly (if we're okay coupling that), or we introduce new autocmd events like `FileRename` and `FileDelete` and LSP subscribes to them.

If we introduce fs events (`FileRename`, `FileDelete`) does that have too much overlap with a "filewatcher" interface (which is on the roadmap but not really designed yet)? I suppose the fs events themselves could be "the filewatcher interface" assuming that gives users enough control to tune things (debounce, directory recursion, etc).

**Comment by tomtomjhj** [comment]
Does this PR subsume #24975 ?

**Comment by stevearc** [comment]
I don't believe that applying workspace edits requires any interaction with `will*`/`did*`. Since the `will*` calls return a workspace edit, if we called it again when applying the workspace edit it would infinite loop.

I personally feel like adding a higher-level API is jumping the gun. We _know_ that file plugin authors need this functionality, because it's been implemented several times over. But those plugins are already handling the filesystem operations themselves. It's not clear to me that there's need or demand for lua APIs that perform create/delete/rename operations (and it might cause more confusion, because there's already vimscript functions, libuv, and `os`). This is also something that could easily be added later, potentially once we better know the shape of the filewatcher interface.

If we _do_ think that this needs to be paired with some higher-level API or event interface, I think it's very important that it support _bulk_ operations. Every call to `willRenameFiles` is a RPC round trip, so we should only do that _once_ with all of the files that we plan to create/delete/rename.

**Comment by mfussenegger** [comment]
> It's not clear to me that there's need or demand for lua APIs that perform create/delete/rename operations (and it might cause more confusion

I think https://github.com/tpope/vim-eunuch shows that there's some demand for functionality like delete and move without pulling in a full file manager. I have a `Del ` command definition myself, and would've a rename for the LSP integration if jdtls didn't provide another way to do that.


> This is also something that could easily be added later, potentially once we better know the shape of the filewatcher interface.

If we add a high level API, then the API introduced here is probably no longer needed, no?
Having both seems too much to me. And if we have to choose between high level functions that users can use directly, or mid-level API that is only useful in combination with some other code I do have a pretty clear preference

**Comment by stevearc** [comment]
> I think https://github.com/tpope/vim-eunuch shows that there's some demand for functionality like delete and move without pulling in a full file manager.

That's a good point! It might make sense to start pulling in some of those commands to core Neovim, and if that happens it would also make sense to wire those up to the LSP client. In this world we would still have this API present, it would just be private. While I don't need this personally, I'm not against it and I can see the value. 

My question is: is there any reason we don't want to expose the low level of the API? We expose `vim.lsp.util.jump_to_location` and `vim.lsp.util.apply_workspace_edit` and many others that interact directly with raw LSP concepts, and that has been enormously helpful for personal configs and plugins alike. This seems conceptually similar.

For me personally, I don't want to use a higher level API for this. If I can just call the LSP methods directly that gives me complete control over when each step is performed (LSP willRename request, file operation, LSP didRename notification) and _how_ each step is performed. I can decide if I want to batch the LSP requests or not. I have full control over the individual file actions (sync vs async, the mode that files/directories are created with, maybe I want to create temp backup files if I'm renaming across devices, maybe I want renames to fail if the target already exists, etc). Maybe I want to confirm every individual workspace edit before applying them, or maybe I want to aggregate them all and confirm the total set, or maybe no confirmation is necessary. There's enough room here for people to want different things that I think that *hiding* the underlying API is going to be counterproductive.

**Comment by mfussenegger** [comment]
> My question is: is there any reason we don't want to expose the low level of the API? We expose vim.lsp.util.jump_to_location and vim.lsp.util.apply_workspace_edit and many others that interact directly with raw LSP concepts, and that has been enormously helpful for personal configs and plugins alike. This seems conceptually similar.

There is no general rule, but some of the factors:

- How general and composable are the functions
- How deep/shallow are they
- Are they used for other built-in functionality

We also consider a few things that were added specifically to vim.lsp.util as a mistake and raised the bar a bit due to that. Some were just in the wrong location, but some functions had/have a problematic design. (See https://github.com/neovim/neovim/issues/25272)

Low level API is in general okay, my reluctance here comes a bit from the facts that:

- It's not really that low level, but somewhere in-between. Low level would be more general, but this targets one rather specific use-case.
- It currently wouldn't be used by anything else in nvim. It would only be there for plugin authors
- Related to ^, we enable new LSP capabilities without really having them built-in/used.


The suggestion/question to go with higher level APIs comes a bit from the fact that there's in general a bit of a push to keep the API surface small if possible.
So far I had the impression that the strategy is often to:

- Expose (very) low level to let people go crazy, and enable new use cases (like the extmarks/virtual_text APIs)
- Expose high level API/commands for end-users
    - Add opts as far as reasonable for customization

Exposing more of the building blocks for the high level end-user functions is more of a last resort, if the low-level API requires too much plumbing, and if extending the opts would be too unflexible.

> I have full control over the individual file actions (sync vs async, the mode that files/directories are created with, maybe I want to create temp backup files if I'm renaming across devices, maybe I want renames to fail if the target already exists, etc). Maybe I want to confirm every individual workspace edit before applying them, or maybe I want to aggregate them all and confirm the total set, or maybe no confirmation is necessary

Fair points. With oil.nvim in mind, how do your concrete functions behave?

---

> If we introduce fs events (FileRename, FileDelete) does that have too much overlap with a "filewatcher" interface (which is on the roadmap but not really designed yet)? I suppose the fs events themselves could be "the filewatcher interface" assuming that gives users enough control to tune things (debounce, directory recursion, etc).

That's a good point. I haven't really thought about that, but yes, I think there would be overlap.

---

Also tagging @MariaSolOs for some more opinions.

**Comment by stevearc** [comment]
> * It currently wouldn't be used by anything else in nvim. It would only be there for plugin authors
> * Related to ^, we enable new LSP capabilities without really having them built-in/used.

These are both very fair. The initial reason for this PR was to expose the glob matching logic via a non-matching API, as that was the only piece that I really didn't want to copy/implement myself. From there, it seemed like there would be some value in putting the rest of the implementation into Neovim core so we don't have to keep re-inventing it. While the API still feels relatively low-level to me, the concern that there's no immediate use for it in core makes a lot of sense.

Maybe the answer here is that this logic _could_ be useful, but it's not the right time for it. Maybe it would make more sense to wait until we add more file manipulation capabilities to core, and at that point it could be worth thinking about it again.

If #26519 or something like it gets merged, then that solves my immediate need.

**Comment by GitMurf** [comment]
sorry it was hard for me to follow the outcome here. Is there a recommended workaround / path for now? it sounds like it was decided to wait for some more core file operations to be implemented into neovim core before pursuing this? In the typescript world this is a huge feature for me. Thanks for everyone's hard work!

**Comment by GitMurf** [comment]
Is it fair to say the best option short term is to use Oil based on this? https://github.com/stevearc/oil.nvim/pull/264

**Comment by stevearc** [comment]
@GitMurf the outcome is that we don't want to upstream these wrappers around the raw LSP methods until there is something in Neovim core that would actually make use of them. For the time being the action should be purely upon plugin authors, who will have to implement this logic themselves. The good news is I think most of the popular file managers already either support these operations out of the box or have another third-party plugin available to add support for them.
